### ISSUE TITLE
[ServiceBus] Observing number of user errors > 0 in metrics in Azure Portal

### ISSUE THREAD

**Issue by yunhaoling** [Azure/azure-sdk-for-python#10741]
- **Package Name**: azure-servicebus
- **Package Version**: 0.50.2
- **Operating System**: Windows
- **Python Version**:  Python3.7

**Describe the bug**
Azure Portal/ServiceBus Namespace/Monitoring/Metrics/User errors observe error number.

**To Reproduce**
1. Create a session enabled queue
2. Used the code bellow (tweak the session pool sample code to not send)
```python
# ------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for
# license information.
# -------------------------------------------------------------------------
import asyncio
import uuid
import conftest
from azure.servicebus.aio import ServiceBusClient, Message
from azure.servicebus.common.constants import NEXT_AVAILABLE
from azure.servicebus.common.errors import NoActiveSession

async def message_processing(queue_client):
    while True:
        try:
            async with queue_client.get_receiver(session=NEXT_AVAILABLE, idle_timeout=1) as session:
                await session.set_session_state("OPEN")
                async for message in session:
                    print("Message: {}".format(message))
                    if str(message) == 'shutdown':
                        await session.set_session_state("CLOSED")
                        break
                    await message.complete()
        except NoActiveSession:
            print("NoActiveSession")
            return

async def sample_session_send_receive_with_pool_async(sb_config, queue):
    concurrent_receivers = 50
    sessions = [str(uuid.uuid4()) for i in range(concurrent_receivers)]
    client = ServiceBusClient(
        service_namespace=sb_config['hostname'],
        shared_access_key_name=sb_config['key_name'],
        shared_access_key_value=sb_config['access_key'])
    queue_client = client.get_queue(queue)
    # for session_id in sessions:
    #     async with queue_client.get_sender(session=session_id) as sender:
    #         await asyncio.gather(*[sender.send(Message("Sample message no. {}".format(i))) for i in range(20)])
    #         await sender.send(Message("shutdown"))
    receive_sessions = [message_processing(queue_client) for _ in range(concurrent_receivers)]
    await asyncio.gather(*receive_sessions)

if __name__ == '__main__':
    while True:
        live_config = conftest.get_live_servicebus_config()
        queue_name = conftest.create_session_queue(live_config)
        loop = asyncio.get_event_loop()
        try:
            loop.run_until_complete(sample_session_send_receive_with_pool_async(live_config, queue_name))
        finally:
            print("complete")
            conftest.cleanup_queue(live_config, queue_name)
```
3. Go to azure portal
    1) Go to the servicebus namespace
    2) Go section **Monitoring**, then **Metrics**
    3) On the right metrics panel, select **User errors** in the drop-down menu under "Metrics" column.
    4) Should be able to see errors

**Expected behavior**
Count of "User error" being 0 in the metrics

**Additional context**
This is an issue found by @MichaelWeedon, the patch addressing the dead_letter fix: https://github.com/Azure/azure-sdk-for-python/issues/9242 seems to solve the problem here

**Comment by yunhaoling** [comment]
Follow up:

the 0.50.3b1 patch doesn't fix the problem. Use a newly-created session-enabled queue can reproduce the problem with 0.50.3b1.

User errors are defined here:
https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-metrics-azure-monitor#user-errors

**Comment by yunhaoling** [comment]
Hi @MichaelWeedon, after investigating the code, I think the use error in metrics in azure portal is  explainable when no session is available because error dues occur. Here is what's happening:

1) NEXT_AVAILABLE is a special setting when creating the connection to service side, asking service side to give client any available session.
2) If there's no session available, the service side would **return error to the client side "uamqp.errors.VendorLinkDetach: com.microsoft:timeout: The operation did not complete within the allocated time 00:01:00 for object broker-linkxxxxxx."**
  - I think this is what user errors is in the metrics in portal, it's expected behavior.
3) The raw error is then wrapped into "NoActiveSession" error and raised out finally

I have checked that .Net SDK would similarly have user error in the metric with newly created session enabled queue but no available session on that queue.

---------

TL;DR: User error on metric is an expected outcome as no available session indeed triggers error.

**Comment by MichaelWeedon** [comment]
It is unclear why this condition is being treated as an exception and tracked as an error.  Being able to process sessions without know the session id is a requirement. Why is this an exception when there are no messages/sessions?

**Comment by yunhaoling** [comment]
Hello @MichaelWeedon , I'm unable to answer further. From my perspective, this a service defined behavior.

The service side need to tell client side that there's no session available, and I believe erroring is the underlying mechanism for doing such.

Hi, @nemakam, would you be able to confirm my understanding is correct that no-available session triggers error on the service side and logs user error in the error metric?

**Comment by nemakam** [comment]
Yes. When you request for a session and no session exists, it is marked as client error. 
However, when you request for a message, and you don't receive a message, it is not an error, and is considered a successful request.

**Comment by MichaelWeedon** [comment]
@nemakam, In a session based scenario, The consumer does not know its session id.  Thus it requests the next available session to see if there are any available. How is that different then requesting a message and none are received?  
That said, What would be the recommended flow/logic for a session based message queue where the session id's are unknown to the consumer such that error are not generated?

**Comment by nemakam** [comment]
@MichaelWeedon , there is no recommendation. There are options - 
1. You could figure out in your pipeline how many sessions are present.
2. You could do a `GetMessageSessions()` or its equivalent in this SDK to figure out if you have any sessions in the queue and then accept the session.
3. You could ignore this user error.
4. Keep track of your errors in your own metrics pipeline and mark this exception as "not an error"

**Comment by MichaelWeedon** [comment]
@nemakam, 
We are trying to ignore the error by catching/handling the exception.
Additionally, the python sample code provided for service bus was used to build our application. This sample code was provided for replicating the problem.  
We ask that the exception not be logged as a User error.
And/or the sample code be updated to reflect the proper way to process a session based Service Bus queue using NEXT_AVAILABLE.

**Comment by nemakam** [comment]
@MichaelWeedon , thanks for the feedback. We will discuss internally to see if and how to proceed with the scenario here. We haven't received any feedback regarding this scenario producing an user error, and we will have to consider back-compat when taking decisions.
Till then, I'd recommend you to ignore the user-error count in this case.

**Comment by yunhaoling** [comment]
Hey @MichaelWeedon , I know you're shifting to our latest SDK v7 but still suffering from this problem.

I see your last request for the SDK sample update:

`the sample code be updated to reflect the proper way to process a session based Service Bus queue using NEXT_AVAILABLE.`

I'd like to know more about your use case:

- Are you mainly using the [session_pool_receive](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/servicebus/azure-servicebus/samples/sync_samples/session_pool_receive.py) to receive messages from sessions?
  - In this case, I think the "proper way" is letting the code keep retrying to get an available session:
```python
while True:
    try:
        with sb_client.get_queue_session_receiver(queue_name, idle_timeout=1) as receiver:
            # receive and handle message
    except NoActiveSession:
        pass
        # you can do logging here
        # entering next loop, which would continue trying to get an available session
```

- If you focus on using a single session receiver, I think the strategy is the same:
```python
receiver = sb_client.get_queue_session_receiver(queue_name, idle_timeout=1)
work_flag = True
while work_flag:
    try:
        with receiver:
            # receive and handle messages
            # set work_flag = False if needs to stop the program
    except NoActiveSession:
        # try to create a new session receiver
        receiver = sb_client.get_queue_session_receiver(queue_name, idle_timeout=1)
```

Does this make sense to you? or do you have any ideas on the sample improvement?

**Comment by MichaelWeedon** [comment]
@yunhaoling, 
Our use case is an unknown number of sessions with unknown session ids, thus the original use of NEXT_AVAILABLE. Using NEXT_AVAILABLE is not a requirement.

The v7.0.0 Sample no longer uses NEXT_AVAILABLE - which is fine. However it still suffers from the Users Errors seen in the Azure Portal Metrics.  

I tried a different approach to receive messages for a specific session by attempting to retrieve the list of sessions first and looping through them.  Since we have a sentinel at the end of each session, we know when to exit the receive loop without timing out.

That said, neither the 0.50.3 not the 7.0.0b2 SDK's have a method for retrieving a proper list of sessions.  In 0.50.3, the list_sessions() an empty list unless the sessions that have had their status set. But setting the status means they are being processed. We want to get a list of the sessions that have not been processed - meaning no status. NOTE: Once the status is set, sessions stick around for a long undetermined amount of time and there is no way in python to clear the status. 

In 7.0.0b2, I could not find a method for getting a list of sessions. 
There is a C# application called Service Bus Explorer that is able to properly retrieve the list of sessions using GetMessageSession(), which BTW returns more information than the session id.

Also, in v7.0.0b2 there are now methods for handling/processing the deadletter queue - insight into how to do this would be appreciated. I could not find a way to report this/ask this question.

Handling the NoActiveSession exception does not prevent the User Errors from appearing in the Portal. 

The goal of all this is to process session based messages with unknown session ids without getting User Errors in the Portal from a python application service.

**Comment by yunhaoling** [comment]
@MichaelWeedon thanks again for your patience and clarifying your scenario.

1) As this is a service behavior, v7 is expected to suffer from this as long as the service still classifies this as a user error. (in fact v7 is using NEXT_AVAILABLE internally if no session id is provided)

    **NEXT STEP:** I have opened a separate issue to draw service team's attention so that they have a better place for tracking and discussion: https://github.com/Azure/azure-sdk-for-python/issues/11690, I've reached out to them and will keep in contact with them to see how we can solve this.

2) As we have limited resources, we're focusing our efforts on champion scenarios for GA. Right now list sessions is supposed to a post-GA feature and it depends on how we prioritize. (Thanks for your information on the list session feature, it's helpful)

3) Dead letter receiver support will come in v7.0.0b3

**Comment by yunhaoling** [comment]
I'm closing this one now, let's use the https://github.com/Azure/azure-sdk-for-python/issues/11690 to keep tracking the service side issue.
